ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/plugins/api/ApplicationPluginBuilder.kt

```diff
@@ -82,151 +50,175 @@ public abstract class PluginBuilder<PluginConfig : Any> internal constructor(
-     * Specifies how to modify sending data within an HTTP call for the current [PluginBuilder].
-     * @see OnCallRespond
+     * Defines how current [ApplicationPluginBuilder] needs to transform data received from a client.
+     *
+     * @param block An action that needs to be executed when your application receives data from a client.
      **/
-    public override val onCallRespond: OnCallRespond = object : OnCallRespond {
-        private val plugin = this@PluginBuilder
-
-        override fun invoke(block: suspend CallRespondContext.(ApplicationCall) -> Unit) {
-            plugin.onDefaultPhase(
-                plugin.onResponseInterceptions,
-                ApplicationSendPipeline.Transform,
-                PHASE_ON_CALL_RESPOND,
-                ::CallRespondContext,
-                block
-            )
-        }
+    public override fun onCallReceive(block: suspend CallReceiveContext<PluginConfig>.(ApplicationCall) -> Unit) {
+        onDefaultPhase(
+            onReceiveInterceptions,
+            ApplicationReceivePipeline.Transform,
+            PHASE_ON_CALL_RECEIVE,
+            ::CallReceiveContext,
+        ) { call, _ -> block(call) }
+    }
 
-        override fun afterTransform(block: suspend CallRespondAfterTransformContext.(ApplicationCall, Any) -> Unit) {
-            plugin.onDefaultPhaseWithMessage(
-                plugin.afterResponseInterceptions,
-                ApplicationSendPipeline.After,
-                PHASE_ON_CALL_RESPOND_AFTER,
-                ::CallRespondAfterTransformContext,
-                block
-            )
-        }
+    /**
+     * Specifies how to transform the data. For example, you can write a custom serializer using this method.
+     *
+     * @param block An action that needs to be executed when your server is sending a response to a client.
+     **/
+    public override fun onCallRespond(
+        block: suspend CallRespondContext<PluginConfig>.(call: ApplicationCall, body: Any) -> Unit
+    ) {
+        onDefaultPhase(
+            onResponseInterceptions,
+            ApplicationSendPipeline.Transform,
+            PHASE_ON_CALL_RESPOND,
+            ::CallRespondContext,
+            block
+        )
     }
 
     /**
      * Executes specific actions after all [targetPlugins] are executed.
      *
-     * @param targetPlugins Plugins that need to be executed before your current [PluginBuilder].
+     * @param targetPlugins Plugins that need to be executed before your current [ApplicationPluginBuilder].
      * @param build Defines the code of your plugin that needs to be executed after [targetPlugins].
      *
      * Note: you can define multiple actions inside a [build] callback for multiple stages of handling an HTTP call
      * (such as [onCall], [onCallRespond], and so on). These actions are executed right after all actions defined
      * by the given [plugin] are already executed in the same stage.
      **/
-    public fun afterPlugins(
+    public fun after(
```

Comment:
I'm not sure about `after` name cuz it does not give you much information. Especially taking in account that now you have `after/*plugins*/(....)` and `on/*event*/()` and also we may introduce `after/*event*/(...)` at some point.

But I would not insist too much on naming here. Just please add it to a list of things that we need to discuss with the whole team

Reply:
we have `after` only for plugins and we had this discussion at the design meeting. @rsinukov wdyt?